src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs

```diff
@@ -98,51 +126,53 @@ private static void DebugVerifyTags(NormalizedSnapshotSpanCollection requestedSp
             }
         }
 
-        private IEnumerable<ITagSpan<TTag>> GetIntersectingTagSpansWorker(NormalizedSnapshotSpanCollection requestedSpans)
+        private void AddIntersectingTagSpansWorker(
+            NormalizedSnapshotSpanCollection requestedSpans,
+            SegmentedList<ITagSpan<TTag>> tags)
         {
             const int MaxNumberOfRequestedSpans = 100;
 
             // Special case the case where there is only one requested span.  In that case, we don't
             // need to allocate any intermediate collections
-            return requestedSpans.Count == 1
-                ? GetIntersectingSpans(requestedSpans[0])
-                : requestedSpans.Count < MaxNumberOfRequestedSpans
-                    ? GetTagsForSmallNumberOfSpans(requestedSpans)
-                    : GetTagsForLargeNumberOfSpans(requestedSpans);
+            if (requestedSpans.Count == 1)
+                AddIntersectingSpans(requestedSpans[0], tags);
+            else if (requestedSpans.Count < MaxNumberOfRequestedSpans)
+                AddTagsForSmallNumberOfSpans(requestedSpans, tags);
+            else
+                AddTagsForLargeNumberOfSpans(requestedSpans, tags);
         }
 
-        private IEnumerable<ITagSpan<TTag>> GetTagsForSmallNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans)
+        private void AddTagsForSmallNumberOfSpans(
+            NormalizedSnapshotSpanCollection requestedSpans,
+            SegmentedList<ITagSpan<TTag>> tags)
         {
-            var result = new List<ITagSpan<TTag>>();
-
-            foreach (var s in requestedSpans)
-            {
-                result.AddRange(GetIntersectingSpans(s));
-            }
-
-            return result;
+            foreach (var span in requestedSpans)
+                AddIntersectingSpans(span, tags);
         }
 
-        private IEnumerable<ITagSpan<TTag>> GetTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans)
+        private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans, SegmentedList<ITagSpan<TTag>> tags)
```

Comment:
note: i don't really understand waht this function is doing.  so all i did was preserve semantics, while making it used pooled temporary buffers.

it seems to be some sort of scratch space where it tries to walk a larger span, but then skip over tags in the gaps, as well as use a hashset to ensure nto duplicates.

i imagine it can be very much cleaned up, but i didn't want to break something subtle int he logic here.